proc.c

```diff
@@ -2695,6 +2695,83 @@ rb_method_curry(int argc, const VALUE *argv, VALUE self)
     return proc_curry(argc, argv, proc);
 }
 
+static VALUE
+compose(VALUE dummy, VALUE args, int argc, VALUE *argv, VALUE passed_proc)
+{
+    VALUE f, g, fargs;
+    f = RARRAY_AREF(args, 0);
+    g = RARRAY_AREF(args, 1);
+    fargs = rb_ary_new3(1, rb_proc_call_with_block(g, argc, argv, passed_proc));
```

Comment:
If we could change this to send `call` to `g` with the given arguments (e.g. via `rb_funcall`), this would work with any object with a `call` method (including other `Proc`s, `Method`s and regular Ruby classes).

Alternatively, we could ensure `g` is a `Proc` by calling `to_proc` on it if it is not already a `Proc` (not entirely dissimilar to using the existing `StringValue` macro).